File: my-folder/0907-koko-eating-bananas/solution.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int n = piles.size(); 
        int high = *max_element(piles.begin(), piles.end());

        if(n == h) return high;
        int low = 1;
        while(low < high){
            int totalHours = 0;
            int mid = low + ((high - low) / 2);
            for(int k = 0; k < n; ++k)
                totalHours += ceil((double)piles[k] / mid);
            
            if (totalHours > h) low = mid + 1;
            else high = mid;
        }
        return low;
    }
};

File: my-folder/0907-koko-eating-bananas/solution.java
class Solution {
    boolean solve(int[] piles,int n,int val, int h){
        int result = 0;
        for(int i=0;i<n;i++){
            if(piles[i]%val!=0) result+=piles[i]/val+1;
            else result+=piles[i]/val;
            if(result>h) return false;
        }
        return true;
        
    }
    public int minEatingSpeed(int[] piles, int h) {
        int n = piles.length;
        // Arrays.sort(piles);
        int left = 1;
        int right = Arrays.stream(piles).max().getAsInt();
        int ans = right;
        while(left<=right){
            int mid = left+(right-left)/2;
            if(solve(piles,n,mid,h)){
                ans = Math.min(mid,ans);
                right=mid-1;
            }
            else left = mid+1;
        }
        return ans;
    }
}
